### DISCUSSION TITLE
Create DAG with Airflow REST API

### DISCUSSION THREAD

**Discussion by luiztauffer** [apache/airflow#24539] [General] [answered]
### Description

Given that all DAGs are serialized in the database ([reference](https://airflow.apache.org/docs/apache-airflow/stable/dag-serialization.html)), it feels like a POST request in the [Airflow REST API](https://airflow.apache.org/docs/apache-airflow/stable/stable-rest-api-ref.html) for creating DAGs would be a very useful improvement! It could accept already serialized dags, for example.

Is this something that might be considered for an upcoming version?

Thanks for all the great work, I’ve been loving working with Airflow!

### Use case/motivation

My context: I’m working on a project where I start with JSON files defining newly created DAGs, and then I see myself having to turn those into Python files (I use Jinja2 templates for that) and having to upload these newly created .py files to a storage somewhere, where a running instance of Airflow can read the new DAGs. 

The above approach works, but certainly things could be better. The necessity of actual python files stored somewhere unnecessarily increases the complexity of each deployment.

I would also appreciate other ideas to possibly circumvent this limitation.

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by vinicvaz** [comment]
following

**Comment by potiuk** [comment]
The serialization only stores dag structure and configuration not python code. So your basic assumption here does not hold unfortunately because what your serialization idea is, it is different that the serialization we have in Airflow.

Python DAG is much more than just a single python file - it is the byte code of the Python file but also potentially imported classes and even potentially files that are read by the Python code placed in the dag folder and any of the subfolders. Potentially loaded in dynamic way by your python code based on multiple factors. If you have 1000s of dag files and 100s of common utilities, you basically have no idea (in generic case) which parts of ti should be serialized). And you generally you might end up with serializing the whole DAG folder which is .... an exact equivalent of what we have when we use "remote filesystems" or "git-sync" - only that currently the remote filesystem takes care about distributing all of it, and in case of "DAG submission" you would have to package it together and send via REST API.

There area number of discussions on how to do what you want resulting in a number of ideas like DAG Fetcher https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=109445755 or DAG Manifest https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-20+DAG+manifest and it involves changing the way and limiting how DAGs are being written and potentially annotated (and serialized including all the information necessary to not only know the DAG structure but also to execute it). 

A lot of the challenges and problems to solve have been very nicely described in the recent talk from the Airflew Summit 2022 (3 weeks ago) where AirBnB explained how they have done something similar internally and what kind of limitations they had to introduce and how much they had to enforce internally to make it happen: [On Demand DAG through the REST API](https://youtu.be/5Ap2t9qJE18)

This is PRECISELY about making Dags submittable via rest API. I recommend anyone who wants to take part further in the discussion to watch that talk as it explains everything you need to know to understand complexities involved.

Speaking ilof which - once you or anyone else who wishes to discuss it watches the talk, think about those issues and understands complexities involved, you are free to propose, discuss (on devlist) an AIP for that. Like everyone in the community you are free to do it and start devlist discussion. 

Just be prepared that you have to lead and get consensus on how to solve those issues in a generic way (Airflow is a platform for everyone, not a solution for specific users only). People will have different opinions and ideas there, and the biggest complexity of the approach will be to get to consensus   - what trade-offs to make, how to implement serialization, how to handle dependencies, how to handle different python versions (pickle!). Most of the issues (without proposing solutions) are mentioned in the talk. And most of the solutions implemented by AirBnB was to heavily limit their use case and enforce their internal limits on how the DAGs are written and distributed - in a way that is specific to AirBnB. 

I personally think it would be possible to have it, but it will be months of discussions at least and years of implementation in incremental steps to get there. And I am not sure if it's worth at all (see the second comment below where I thought a bit more about this).

So just be prepared for a marathon not a sprint - whoever wants to lead that discussion.

**Comment by potiuk** [comment]
Converted it to a discussion as this is far beyond a "feature"

**Comment by potiuk** [comment, marked as answer]
Another comment here:

> My context: I’m working on a project where I start with JSON files defining newly created DAGs, and then I see myself having to turn those into Python files (I use Jinja2 templates for that) and having to upload these newly created .py files to a storage somewhere, where a running instance of Airflow can read the new DAGs.

> The above approach works, but certainly things could be better. The necessity of actual python files stored somewhere unnecessarily increases the complexity of each deployment.

Actually this is a very good approach.  Airflow as a platform is designed in the way to take "Python" as DAG definition. We will never turn into "declarative" way of writing DAGs because this will severly limit the use cases that the "Airflow Platform" will handle. As a platform we need to be generic and versatile enough to handle multiple cases.  

And different users are perfectly free to limit their cases to smaller scope and write their DAGs in declarative way. Converting JSON or other declarative forms of DAGS to Python is perfectly OK  (and JInja is pretty good at that - been there done that and this is working surprisingly well and is very simple to manage and maintain). Also another approach where you combine manually written Python code and some data files (JSON for example) are possible and nicely [described in our docs](https://airflow.apache.org/docs/apache-airflow/stable/howto/dynamic-dag-generation.html?highlight=dynamic%20dags).

Also "filesystem" submission interface rather than REST API is very good as well. What's wrong with it? Instead of making HTTP request you copy files using the filesystem of your choice, Many of those filesystems have even good APIs that can be used for it. 

But you can do better than that even now, and this actually what I HIGHLY recommend. You can use `git` as in intermediary medium and use `git-sync` instead of remote filesystems. 

This has multiple advantages:

* you can keep generated files in `git` in an organized way (Git is great storage for source code)
* you can see the differences and track history
* you can limit access to the repository
* you can bring together muiltiple repositories from different teams via submodules
* you can use pull-request/approval workflow og Git to manage reviews and deployment workflows
* you can even connect it to CI/CD to perform automated validation and testing of the DAGs

There are many talks at the Airflow Summit 2022 -  where our users not only expressed how important those good engineering practices for DAG generation and distribution but also they have shown how they did it:

* [Manage Dags at scale: Dags versioning & package management
](https://www.youtube.com/watch?v=uA-8Lj1RNgw&list=PLGudixcDaxY2LxjeHpZRtzq7miykjjFOn&index=4)
* [Happy DAGs Happy Teammates How a little CICD can go a long way
](https://www.youtube.com/watch?v=Wf5VNH9su9I&list=PLGudixcDaxY2LxjeHpZRtzq7miykjjFOn&index=12)
* [git push your data stack with Airbyte, Airflow and dbt
](https://www.youtube.com/watch?v=_pLDo04sv2U&list=PLGudixcDaxY2LxjeHpZRtzq7miykjjFOn&index=19)

So there are multiple advantages of using Git as an intermediary and GitSync on the Airlfow side comparing to `submit a DAG via REST API`. This solves many, many more problems of yours that REST API submission of DAGs does not even touch on and you will have to solve them anyway. 

I think personally the value of possibly-but-not-likely-but-for-sure-not-near-future `submitting the DAGS via REST API` is grossly overrated comparing to the value of other ways that you can use even today to make your Airlfow DAG submission experience easier. It seems like "natural" way to do it - to submit DAGs via REST API, but in-fact the complexity it brings comparing to the value it has, makes it very low priority in general - especially that there are better ways to handle this already today - with many more advantages and with similar "user" complexity.

**Comment by luiztauffer** [comment]
Thank you for your detailed answers, @potiuk 

Indeed the idea of using git as a storage solution for DAGs sounds better than what I had in mind before! This would be beautiful for versioning and some extra functionalities I wasn’t actually considering before

cheers!
